src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs

```diff
@@ -1630,6 +1654,63 @@ private void MakeSeparatorList(ReadOnlySpan<char> separators, ref ValueListBuild
             }
         }
 
+        private void MakeSeparatorListVectorized(ref ValueListBuilder<int> sepListBuilder, char c, char? c2 = null, char? c3 = null)
+        {
+            // Constant that defines indices of characters within an AVX-Register
+            const ulong indicesConstant = 0xFEDCBA9876543210;
+            // Constant that allows for the truncation of 16-bit (FFFF/0000) values within a register to 4-bit (F/0)
+            Vector256<byte> shuffleConstant = Vector256.Create(0x02, 0x06, 0x0A, 0x0E, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+                                                               0xFF, 0xFF, 0xFF, 0xFF, 0x02, 0x06, 0x0A, 0x0E, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF);
+
+            Vector256<ushort> v1 = Vector256.Create(c);
+            Vector256<ushort> v2 = c2 is char sep2 ? Vector256.Create(sep2) : v1;
+            Vector256<ushort> v3 = c3 is char sep3 ? Vector256.Create(sep3) : v2;
+
+            ref char c0 = ref MemoryMarshal.GetReference(this.AsSpan());
+            int cond = Length - (Length % Vector256<ushort>.Count);
+            int i = 0;
+
+            for (; i < cond; i += Vector256<ushort>.Count)
+            {
+                Vector256<ushort> charVector = ReadVector(ref c0, i);
+                Vector256<ushort> cmp = Avx2.CompareEqual(charVector, v1);
+
+                cmp = Avx2.Or(Avx2.CompareEqual(charVector, v2), cmp);
+                cmp = Avx2.Or(Avx2.CompareEqual(charVector, v3), cmp);
+
+                if (Avx.TestZ(cmp, cmp)) { continue; }
+
+                Vector256<byte> mask = Avx2.ShiftLeftLogical(cmp.AsUInt64(), 4).AsByte();
+                mask = Avx2.Shuffle(mask, shuffleConstant);
+
+                Vector128<byte> res = Sse2.Or(mask.GetLower(), mask.GetUpper());
+                ulong extractedBits = Bmi2.X64.ParallelBitExtract(indicesConstant, Sse2.X64.ConvertToUInt64(res.AsUInt64()));
+
+                while (true)
+                {
+                    sepListBuilder.Append(((byte)(extractedBits & 0xF)) + i);
```

Comment:
This causes the vectors to be spilled to the stack.

<details>
  <summary>asm from loop</summary>

```asm
G_M15376_IG06:
       8BFA                 mov      edi, edx
       C4C17E6F247F         vmovdqu  ymm4, ymmword ptr[r15+2*rdi]
       C5FD114D90           vmovupd  ymmword ptr[rbp-70H], ymm1
       C5DD75E9             vpcmpeqw ymm5, ymm4, ymm1
       C5FD119570FFFFFF     vmovupd  ymmword ptr[rbp-90H], ymm2
       C5DD75F2             vpcmpeqw ymm6, ymm4, ymm2
       C5CDEBED             vpor     ymm5, ymm6, ymm5
       C5FD119D50FFFFFF     vmovupd  ymmword ptr[rbp-B0H], ymm3
       C5DD75E3             vpcmpeqw ymm4, ymm4, ymm3
       C5DDEBED             vpor     ymm5, ymm4, ymm5
       C4E27D17ED           vptest   ymm5, ymm5
       0F84B7010000         je       G_M15376_IG22
						;; bbWeight=4    PerfScore 57.67
G_M15376_IG07:
       C5DD73F504           vpsllq   ymm4, ymm5, 4
       C5FD1145B0           vmovupd  ymmword ptr[rbp-50H], ymm0
       C4E25D00E0           vpshufb  ymm4, ymm4, ymm0
       C4E37D39E501         vextracti128 ymm5, ymm4, 1
       C5D9EBE5             vpor     xmm4, xmm4, xmm5
       C4E1F97EE1           vmovd    rcx, xmm4
       49B81032547698BADCFE mov      r8, 0xFEDCBA9876543210
       4933C8               xor      rcx, r8
						;; bbWeight=2    PerfScore 17.67
G_M15376_IG08:
       48898D40FFFFFF       mov      qword ptr [rbp-C0H], rcx
       448BC1               mov      r8d, ecx
       4183E00F             and      r8d, 15
       899548FFFFFF         mov      dword ptr [rbp-B8H], edx
       4403C2               add      r8d, edx
       4489853CFFFFFF       mov      dword ptr [rbp-C4H], r8d
       448B4B08             mov      r9d, dword ptr [rbx+8]
       4C8D5310             lea      r10, bword ptr [rbx+16]
       4C899528FFFFFF       mov      bword ptr [rbp-D8H], r10
       498BFA               mov      rdi, r10
       44898D38FFFFFF       mov      dword ptr [rbp-C8H], r9d
       443B4F08             cmp      r9d, dword ptr [rdi+8]
       7C08                 jl       SHORT G_M15376_IG10
						;; bbWeight=16    PerfScore 184.00
G_M15376_IG09:
       488BFB               mov      rdi, rbx
       E870F1FFFF           call     System.Collections.Generic.ValueListBuilder`1[Int32][System.Int32]:Grow():this
						;; bbWeight=4    PerfScore 5.00
G_M15376_IG10:
       4C8B9528FFFFFF       mov      r10, bword ptr [rbp-D8H]
       448B8D38FFFFFF       mov      r9d, dword ptr [rbp-C8H]
       453B4A08             cmp      r9d, dword ptr [r10+8]
       0F8334010000         jae      G_M15376_IG23
       498B3A               mov      rdi, bword ptr [r10]
       4D63D1               movsxd   r10, r9d
       448B853CFFFFFF       mov      r8d, dword ptr [rbp-C4H]
       46890497             mov      dword ptr [rdi+4*r10], r8d
       41FFC1               inc      r9d
       44894B08             mov      dword ptr [rbx+8], r9d
       488B8D40FFFFFF       mov      rcx, qword ptr [rbp-C0H]
       48C1E904             shr      rcx, 4
       4885C9               test     rcx, rcx
       0F85EC000000         jne      G_M15376_IG21
						;; bbWeight=16    PerfScore 236.00
G_M15376_IG11:
       8B9548FFFFFF         mov      edx, dword ptr [rbp-B8H]
       83C210               add      edx, 16
       8BB54CFFFFFF         mov      esi, dword ptr [rbp-B4H]
       3BD6                 cmp      edx, esi
       89B54CFFFFFF         mov      dword ptr [rbp-B4H], esi
       C5FD1045B0           vmovupd  ymm0, ymmword ptr[rbp-50H]
       C5FD104D90           vmovupd  ymm1, ymmword ptr[rbp-70H]
       C5FD109570FFFFFF     vmovupd  ymm2, ymmword ptr[rbp-90H]
       C5FD109D50FFFFFF     vmovupd  ymm3, ymmword ptr[rbp-B0H]
       0F82D9FEFFFF         jb       G_M15376_IG06
```

It's not exactely the loop-code from this PR, it's with some minor modifications from my attempts to remove the spills. The code from the PR has the same spills, anyway.
</details>

I don't understand why the JIT spills the vectors. They could (and should) be kept in the registers, as the `ValueListBuilder<int>` doesn't need any registers. Or is this from the builder's internal usage of span and zeroing behavior, but even if so, the JIT shouldn't spill the vectors...

Reply:
Does it matter though? The vectors don't seem to be used after L1687 anyways?

Reply:
> Or is this about shuffleConstant, v1, v2, v3?

No.

> Does it matter though? 

Perf-wise: yes. 
* stack-spills are (here) unnecessary memory accesses
* the machine code will be quite a lot smaller without the spills -- especially important as it's within a loop

Reply:
I ended up removing the need for PEXT and just iterated over each potential index within the extracted vector. For some reason I've been seeing some decent performance gains, even in scenarios with low separator frequency (which to me would point to perf regressions over PEXT). Could this inadvertently have fixed the stack-spills? Either way I can't quite understand the perf uplift.